sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -87,6 +87,14 @@ case class DataSource(
   lazy val providingClass: Class[_] = DataSource.lookupDataSource(className)
   lazy val sourceInfo: SourceInfo = sourceSchema()
   private val caseInsensitiveOptions = CaseInsensitiveMap(options)
+  private val equality = sparkSession.sessionState.conf.resolver
+
+  userSpecifiedSchema.foreach { dataSchema =>
+    SchemaUtils.checkColumnNameDuplication(
+      dataSchema.map(_.name), "in the data schema", equality)
+  }
+  SchemaUtils.checkColumnNameDuplication(
+    partitionColumns, "in the partition schema", equality)
 
```

Comment:
@cloud-fan We need to check the duplication for `bucketSpec` here? The current version also could catch the case in `bucketSpec` like;
```
scala> Seq((1, 2)).toDF("a", "b").write.bucketBy(3, "a", "a").saveAsTable("t")
org.apache.spark.sql.AnalysisException: Found duplicate column(s) in the bucket: `a`;
  at org.apache.spark.sql.util.SchemaUtils$.checkColumnNameDuplication(SchemaUtils.scala:85)
  at org.apache.spark.sql.util.SchemaUtils$.checkColumnNameDuplication(SchemaUtils.scala:67)
  at org.apache.spark.sql.execution.datasources.PreprocessTableCreation.normalizeBucketSpec(rules.scala:289)
  at org.apache.spark.sql.execution.datasources.PreprocessTableCreation.org$apache$spark$sql$execution$datasources$PreprocessTableCreation$$normalizeCatalogTable(rules.scala:232)
  at org.apache.spark.sql.execution.datasources.PreprocessTableCreation$$anonfun$apply$2.applyOrElse(rules.scala:197)
  at org.apache.spark.sql.execution.datasources.PreprocessTableCreation$$anonfun$apply$2.applyOrElse(rules.scala:76)
```